File: 프로그래머스/2/86052. 빛의 경로 사이클/빛의 경로 사이클.js
function solution(grid) {
    const answer = [];
    const directions = [[0, 1], [1, 0], [0, -1], [-1, 0]];
    const visit = Array.from({ length: grid.length }, () => 
        Array.from({ length: grid[0].length }, () => Array(4).fill(0))
    );

    function pathfind(r, c, d) {
        let count = 0;
        while (!visit[r][c][d]) {
            visit[r][c][d] = 1;
            count++;
            if (grid[r][c] === 'L') {
                d = (d + 3) % 4; // equivalent to (d - 1) % 4
            } else if (grid[r][c] === 'R') {
                d = (d + 1) % 4;
            }
            r = (r + directions[d][0] + grid.length) % grid.length;
            c = (c + directions[d][1] + grid[0].length) % grid[0].length;
        }
        return count;
    }

    for (let r = 0; r < grid.length; r++) {
        for (let c = 0; c < grid[0].length; c++) {
            for (let d = 0; d < 4; d++) {
                if (visit[r][c][d]) continue;
                answer.push(pathfind(r, c, d));
            }
        }
    }

    return answer.sort((a, b) => a - b);
}

File: 프로그래머스/2/86052. 빛의 경로 사이클/빛의 경로 사이클.py
def solution(grid):
    answer = []
    direc = [[0, 1], [1, 0], [0, -1], [-1, 0]]
    visit = [[[0] * 4 for _ in range(len(grid[0]))] for _ in range(len(grid))]
    
    def pathfind(r, c, d):
        count = 0
        while not visit[r][c][d]:
            visit[r][c][d] = 1
            count += 1 
            if grid[r][c] == 'L': d = (d - 1) % 4
            elif grid[r][c] == 'R': d = (d + 1) % 4
            r = (r + direc[d][0]) % len(grid)
            c = (c + direc[d][1]) % len(grid[0])
            
        return count
    
    for r in range(len(grid)):
        for c in range(len(grid[0])):
            for d in range(4):
                if visit[r][c][d]: continue
                answer.append(pathfind(r, c, d))
                
    return sorted(answer)
